### ISSUE TITLE
fsnotify: build failure on Darwin

### ISSUE THREAD

**Issue by dhess** [NixOS/nixpkgs#17243]
## Issue description

fsnotify fails to build on Darwin due to the old `ld: framework not found Cocoa` issue. I'm building it as a dependency of `stack` with `haskell.packages.ghc801`.

```
Building fsnotify-0.2.1...
Preprocessing library fsnotify-0.2.1...
[1 of 7] Compiling System.FSNotify.Path ( src/System/FSNotify/Path.hs, dist/build/System/FSNotify/Path.o )

src/System/FSNotify/Path.hs:17:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
[2 of 7] Compiling System.FSNotify.Types ( src/System/FSNotify/Types.hs, dist/build/System/FSNotify/Types.o )
[3 of 7] Compiling System.FSNotify.Listener ( src/System/FSNotify/Listener.hs, dist/build/System/FSNotify/Listener.o )
[4 of 7] Compiling System.FSNotify.OSX ( src/System/FSNotify/OSX.hs, dist/build/System/FSNotify/OSX.o )

src/System/FSNotify/OSX.hs:30:34: warning: [-Wunused-top-binds]
    Defined but not used: data constructor ‘Recursive’
[5 of 7] Compiling System.FSNotify.Polling ( src/System/FSNotify/Polling.hs, dist/build/System/FSNotify/Polling.o )

src/System/FSNotify/Polling.hs:14:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
[6 of 7] Compiling System.FSNotify  ( src/System/FSNotify.hs, dist/build/System/FSNotify.o )

src/System/FSNotify.hs:67:1: warning: [-Wunused-imports]
    The import of ‘Control.Applicative’ is redundant
      except perhaps to import instances from ‘Control.Applicative’
    To import instances alone, use: import Control.Applicative()
[7 of 7] Compiling System.FSNotify.Devel ( src/System/FSNotify/Devel.hs, dist/build/System/FSNotify/Devel.o )

src/System/FSNotify/Devel.hs:52:1: warning: [-Wredundant-constraints]
    • Redundant constraint: Monad m
    • In the type signature for:
           doAllEvents :: Monad m => (FilePath -> m ()) -> Event -> m ()
ld: framework not found Cocoa
clang-3.7: error: linker command failed with exit code 1 (use -v to see invocation)
`cc' failed in phase `Linker'. (Exit code: 1)
builder for ‘/nix/store/qa2bprbxav2h2szrqynx3s3sp6xj9645-fsnotify-0.2.1.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/9x3inywzarlnwyajd9zlxwchsfq1xd7s-stack-1.1.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/6k20qv6hr4aqnkq322qzabx21skd2c5s-stack-env.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/6k20qv6hr4aqnkq322qzabx21skd2c5s-stack-env.drv’ failed
```

## Technical details

* System: OS X 10.11.6
* Nix version: `nix-env (Nix) 1.11.2`
* Nixpkgs version: master rev `ac93e9f2c8bbdee7b3c88d8eb358fe4f81fb578a`

See also https://github.com/NixOS/nixpkgs/pull/10605

**Comment by dhess** [comment]
I fixed this in a local branch and now I'm having the problem with `stack` itself, as well.

```
Building stack-1.1.2...
Preprocessing library stack-1.1.2...
[ 1 of 87] Compiling Stack.FileWatch  ( src/Stack/FileWatch.hs, dist/build/Stack/FileWatch.o )
[ 2 of 87] Compiling System.Process.PagerEditor ( src/System/Process/PagerEditor.hs, dist/build/System/Process/PagerEditor.o )
[ 3 of 87] Compiling System.Process.Log ( src/System/Process/Log.hs, dist/build/System/Process/Log.o )
[ 4 of 87] Compiling System.Process.Read ( src/System/Process/Read.hs, dist/build/System/Process/Read.o )
ld: framework not found Cocoa
clang-3.7: error: linker command failed with exit code 1 (use -v to see invocation)
`cc' failed in phase `Linker'. (Exit code: 1)
builder for ‘/nix/store/lr641iyvglm1l9rw95mz49grnmll1ian-stack-1.1.2.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/x8i8s64d0lpvsiy0ij7hvayi5a3fx1gs-stack-env.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/x8i8s64d0lpvsiy0ij7hvayi5a3fx1gs-stack-env.drv’ failed
```

Seems there's some kind of transitive link failure going on as `fsnotify` nor `stack` depend on Cocoa explicitly; I believe it's just `hfsevents` that does.

**Comment by LnL7** [comment]
This was fixed by #17564
